community/security/src/main/java/org/neo4j/server/security/auth/AbstractUserRepository.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.server.security.auth;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.server.security.auth.exception.ConcurrentModificationException;
+
+public abstract class AbstractUserRepository extends LifecycleAdapter implements UserRepository
+{
+    /** Quick lookup of users by name */
+    protected final Map<String, User> usersByName = new ConcurrentHashMap<>();
+
+    /** Master list of users */
+    protected volatile List<User> users = new ArrayList<>();
+
+    private final Pattern usernamePattern = Pattern.compile( "^[a-zA-Z0-9_]+$" );
+
+    @Override
+    public synchronized User getUserByName( String username )
```

Comment:
I missed to comment the first round that this does not need to be synchronized... If we need another rebase.. otherwise leave it for later.